code/game/objects/machinery/vending/vending.dm

```diff
@@ -647,104 +639,119 @@
 		if(item_to_stock.type != R.product_path)
 			continue
 		record = R
 
 	if(!record) //Item isn't listed in the vending records.
-		display_message_and_visuals(user, show_feedback, "[item_to_stock] doesn't belong here!", VENDING_RESTOCK_DENY)
+		user.balloon_alert(user, "[item_to_stock] doesn't belong here!")
 		return FALSE
 
+	return do_stock(item_to_stock, user, show_feedback, record)
+
+///Actually does the restock. Overridden by lasgun vendor for snowflake behaviour
+/obj/machinery/vending/proc/do_stock(obj/item/item_to_stock, mob/user, show_feedback = TRUE, datum/vending_product/record)
+	if(!record.item_can_be_restocked(item_to_stock, user, show_feedback))
+		return FALSE
+	display_message_and_visuals(user, show_feedback, "Restocked", VENDING_RESTOCK_ACCEPT)
+	return TRUE //Item restocked, no reason to go on.
+
+/obj/machinery/vending/lasgun/do_stock(obj/item/item_to_stock, mob/user, show_feedback = TRUE, datum/vending_product/record)
+	//Special snowflake handling of cells
+	///In case of a cell, the amount of charge required to fully charge said cell
+	var/recharge_amount = 0
+	if(iscell(item_to_stock))
+		var/obj/item/cell/cell = item_to_stock
+		if(cell.charge < cell.maxcharge)
+			// Item is not full. Time to try to recharge
+			recharge_amount = cell.maxcharge - cell.charge
+			if(machine_current_charge == 0)
+				display_message_and_visuals(user, show_feedback, "No power!", VENDING_RESTOCK_DENY)
+				return FALSE
+			else if(machine_current_charge < recharge_amount) // Not enough but some charge remaining so partially recharge cell and move on
+				cell.give(machine_current_charge)
+				machine_current_charge = 0
+				cell.update_icon()
+				display_message_and_visuals(user, show_feedback, "Cell charged partially! [round(cell.percent())]%.", VENDING_RESTOCK_RECHARGE)
+				playsound(loc, 'sound/machines/hydraulics_1.ogg', 25, 0, 1)
+				return FALSE
+			else
+				machine_current_charge -= recharge_amount
+				cell.give(recharge_amount)
+				if(!record.item_can_be_restocked(item_to_stock, user, show_feedback))
+					return FALSE
+				display_message_and_visuals(user, show_feedback, "Restocked and recharged", VENDING_RESTOCK_ACCEPT_RECHARGE)
+				return TRUE
+	return ..()
+
+/**
+ * A few checks to make sure the item we are trying to restock is allowed to be restocked
+ * Then restocks it afterwards
+ * Returns TRUE if we successfully restock
+ */
+/datum/vending_product/proc/item_can_be_restocked(obj/item/item_to_stock, mob/user, show_feedback = TRUE)
 	//More accurate comparison between absolute paths.
 	if(isstorage(item_to_stock)) //Nice try, specialists/engis
 		var/obj/item/storage/storage_to_stock = item_to_stock
 		if(!(storage_to_stock.flags_storage & BYPASS_VENDOR_CHECK)) //If your storage has this flag, it can be restocked
-			display_message_and_visuals(user, show_feedback, "Can't restock containers!", VENDING_RESTOCK_DENY)
+			user.balloon_alert(user, "Can't restock containers!")
```

Comment:
user is an optional arg, so these balloon alerts need to be conditional on user actually existing. Did you actually test this?